dali/kernels/common/cast.cuh

```diff
@@ -26,17 +26,48 @@ struct CastSampleDesc {
   const void *input;
 };
 
+struct CastSampleBlockDesc {
+  int first_block;
+  int sample_size;
+};
+
 template <typename OType, typename IType>
-__global__ void BatchedCastKernel(const CastSampleDesc *samples, const BlockDesc<1> *blocks) {
-  const auto &block = blocks[blockIdx.x];
-  const auto &sample = samples[block.sample_idx];
+__device__ __forceinline__ void CastKernelInternal(const CastSampleDesc& sample,
+                                                   int block_start, int block_end) {
   auto *out = static_cast<OType *>(sample.output);
   const auto *in = static_cast<const IType *>(sample.input);
-  for (int x = threadIdx.x + block.start.x; x < block.end.x; x += blockDim.x) {
+  for (int x = threadIdx.x + block_start; x < block_end; x += blockDim.x) {
     out[x] = ConvertSat<OType>(in[x]);
   }
 }
 
+template <typename OType, typename IType>
+__global__ void BatchedCastKernel(const CastSampleDesc *samples, const BlockDesc<1> *blocks) {
+  const auto &block = blocks[blockIdx.x];
+  const auto &sample = samples[block.sample_idx];
+  CastKernelInternal<OType, IType>(sample, block.start.x, block.end.x);
+}
+
+template <typename OType, typename IType>
+__global__ void BinSearchCastKernel(const CastSampleDesc *samples,
```

Comment:
Cool change, I would add a comment in the code, and not only in the PR description, that we store the earliest block that should process given sample and in the given block we do binsearch for its samples.

Reply:
Done